### PR TITLE
Expose cloud provider backoff config 

### LINKED ISSUE #1382: Cloud provider config customization - cloudProviderBackoff
/kind feature

**Describe the solution you'd like**
#1373 allowed rate limits to be customized in cloud provider config. We also might want to do this for other properties. This issue tracks cloud provider backoff config that we want to override. specifically the following ones:

Name | Description | Remark
-- | -- | --
cloudProviderBackoff | Enable exponential backoff to manage resource request retries | Boolean value, default to false
cloudProviderBackoffRetries | Backoff retry limit | Integer value, valid if&nbsp;cloudProviderBackoff&nbsp;is true
cloudProviderBackoffExponent | Backoff exponent | Float value, valid if&nbsp;cloudProviderBackoff&nbsp;is true
cloudProviderBackoffDuration | Backoff duration | Integer value, valid if&nbsp;cloudProviderBackoff&nbsp;is true
cloudProviderBackoffJitter | Backoff jitter | Float value, valid if&nbsp;cloudProviderBackoff&nbsp;is true

**Anything else you would like to add:**
https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This pr allows for customization of cloud provider backoff config through AzureCluster spec.

**Which issue(s) this PR fixes**:
Fixes #1382

**TODOs**:
- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
```release-note
Expose cloud provider backoff config
```

### CODE REVIEW

File: api/v1alpha4/types.go [hunk 1]
```diff
@@ -515,6 +515,26 @@ type AddressRecord struct {
 // CloudProviderConfigOverrides represents the fields that can be overridden in azure cloud provider config.
 type CloudProviderConfigOverrides struct {
 	RateLimits []RateLimitSpec `json:"rateLimits,omitempty"`
+	BackOffs   []BackOffSpec   `json:"backOffs,omitempty"`
```
Comment (shysank) [inline api/v1alpha4/types.go:518]: Do we need `BackOffSpec` here? Unlike rate limits, there is not going to be different backoff configurations. Can we directly use `BackOffConfig` here?

File: controllers/helpers.go [hunk 2]
```diff
@@ -301,6 +301,10 @@ func (cpc *CloudProviderConfig) overrideFromSpec(d azure.ClusterScoper) *CloudPr
 			cpc.AvailabilitySetRateLimit = toCloudProviderRateLimitConfig(rateLimit.Config)
 		}
 	}
+
+	for _, backOff := range d.CloudProviderConfigOverrides().BackOffs {
```
Comment (shysank) [inline controllers/helpers.go:305]: We won't need this loop once the above comment is addressed.

**Review by shysank — COMMENTED** [review]
@Ankitasw Thanks for the PR! Just a couple of comments on the Backoffs data structure, otherwise looks good so far pending tests.

**Comment by nader-ziada** [comment]
looks good

File: controllers/helpers.go [hunk 3]
```diff
@@ -375,6 +378,31 @@ type RateLimitConfig struct {
 	CloudProviderRateLimitBucketWrite int     `json:"cloudProviderRateLimitBucketWrite,omitempty"`
 }
 
+// BackOffConfig indicates the back-off config options.
+// This is a copy of the struct used in cloud-provider-azure: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/d585c2031925b39c925624302f22f8856e29e352/pkg/azureclients/azure_client_config.go#L48
+type BackOffConfig struct {
+	CloudProviderBackoff         bool    `json:"cloudProviderBackoff,omitempty"`
+	CloudProviderBackoffRetries  int     `json:"cloudProviderBackoffRetries,omitempty"`
+	CloudProviderBackoffExponent float32 `json:"cloudProviderBackoffExponent,omitempty"`
```
Comment (shysank) [inline controllers/helpers.go:386]: This should be `float64` to be in line with [cloud-provider-azure](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/d585c2031925b39c925624302f22f8856e29e352/pkg/provider/azure.go#L186)

File: controllers/helpers.go [hunk 4]
```diff
@@ -375,6 +378,31 @@ type RateLimitConfig struct {
...
+	CloudProviderBackoffDuration int     `json:"cloudProviderBackoffDuration,omitempty"`
+	CloudProviderBackoffJitter   float32 `json:"cloudProviderBackoffJitter,omitempty"`
```
Comment (shysank) [inline controllers/helpers.go:388]: Same as above => https://github.com/kubernetes-sigs/cloud-provider-azure/blob/d585c2031925b39c925624302f22f8856e29e352/pkg/provider/azure.go#L188

File: controllers/helpers.go [hunk 5]
```diff
@@ -375,6 +378,31 @@ type RateLimitConfig struct {
...
+}
+
+// toCloudProviderBackOffConfig returns converts infrav1.BackOffConfig to BackOffConfig that is required with the cloud provider.
+func toCloudProviderBackOffConfig(source infrav1.BackOffConfig) *BackOffConfig {
```
Comment (devigned) [inline controllers/helpers.go:392]: Why return a ptr here if when it is used, it is dereferenced?
```go
cpc.BackOffConfig = *toCloudProviderBackOffConfig(d.CloudProviderConfigOverrides().BackOffs)
```

**Review by devigned — CHANGES_REQUESTED** [review]
Consider returning a struct rather than a ptr from `toCloudProviderBackOffConfig`.

**Review by devigned — APPROVED** [review]
lgtm

Thank you!

**Pull request merged**